packages/authentication/test/poc/mock-express-response.js

```diff
@@ -0,0 +1,71 @@
+var http = require('http');
+var mimeTypes = require('./mime-types');
+var res = Object.create(http.ServerResponse.prototype);
+
+var exts = {};
+Object.keys(mimeTypes).forEach(function(key) {
+  var extArray = mimeTypes[key];
+  extArray.forEach(function(ext) {
+    exts[ext] = key;
+  });
+});
+
+res.type = function(t) {
+  return this.setHeader('Content-Type', exts[t]);
+}
+
+res.send = function send(body) {
```

Comment:
Is this code mostly copied from express ([res.send() source](https://github.com/expressjs/express/blob/c087a45b9cc3eb69c777e260ee880758b6e03a40/lib/response.js#L106-L215))? If it is, then we need to give them credit.

@ritch Do you have any vision for how is `res.send()` going to work our DI-based approach `ctx.bind('result', ...)` (which is similar to Koa's `ctx.result`)? Route handlers which call `res.send()` typically don't call `next()`, therefore there is no way how to run "after" hooks.

```js

```

I think we need to implement `res.send` along the following lines:

```js
res.send = function(body) {
  ctx.bind('response.body').toValue(body);
  // get the callback (Promise resolver) for the current handler/ctor function
  const next = ctx.get('next');
  // call it from the next tick to allow the caller of send() to finish first
  setImmediate(next);
});

response.type = function(t) {
  const type = exts[t];
  ctx.bind('response.headers.content-type').toValue(type);
};
```

Thoughts?

cc @raymondfeng @superkhau 

---

Having wrote all of this, I don't understand why do we actually need to provide `res.send` and `res.type` methods? AFAICT, Passport strategies should be using only a very limited subset of `HttpResponse` methods, see https://github.com/jaredhanson/passport-strategy/blob/3d96d4ace0537db091604945b1b4188d9e8aa9c5/README.md#augmented-methods:

 - `.success(user, info)`
 - `.fail(challenge, status)`
 - `.redirect(url, status)`
 - `.pass()`
 - `.error(err)`

Take a look at [passport-oauth2](https://github.com/jaredhanson/passport-oauth2/blob/a02839afb8d15f7d283ae09167973e4d22e8faf8/lib/strategy.js) for an example.